rust/tool/systemd/src/install.rs

```diff
@@ -214,122 +157,148 @@ impl Installer {
-    /// `GenerationArtifacts`. Does not install any files to the ESP.
-    ///
-    /// This function expects an already pre-populated `GenerationArtifacts`. It can only be called
-    /// if ALL unsigned artifacts are already built and stored in `GenerationArtifacts`. More
-    /// specifically, this function can only be called after `build_unsigned_generation_artifacts`
-    /// has been executed.
-    fn build_signed_generation_artifacts(
-        &mut self,
-        generation: &Generation,
-        generation_artifacts: &mut GenerationArtifacts,
-    ) -> Result<()> {
-        let tempdir = &generation_artifacts.tempdir;
-
-        let bootspec = &generation.spec.bootspec.bootspec;
-
-        let esp_gen_paths = EspGenerationPaths::new(&self.esp_paths, generation, self.arch)?;
-
-        let kernel_cmdline =
-            assemble_kernel_cmdline(&bootspec.init, bootspec.kernel_params.clone());
-
+        // Assemble, sign and install the Lanzaboote stub.
         let os_release = OsRelease::from_generation(generation)
             .context("Failed to build OsRelease from generation.")?;
         let os_release_path = tempdir
             .write_secure_file(os_release.to_string().as_bytes())
             .context("Failed to write os-release file.")?;
-
-        let kernel_path: &Path = generation_artifacts
-            .files
-            .get(&esp_gen_paths.kernel)
-            .context("Failed to retrieve kernel path from GenerationArtifacts.")?
-            .into();
-
-        let initrd_path = generation_artifacts
-            .files
-            .get(&esp_gen_paths.initrd)
-            .context("Failed to retrieve initrd path from GenerationArtifacts.")?
-            .into();
-
+        let kernel_cmdline =
+            assemble_kernel_cmdline(&bootspec.init, bootspec.kernel_params.clone());
         let lanzaboote_image = pe::lanzaboote_image(
-            tempdir,
+            &tempdir,
             &self.lanzaboote_stub,
             &os_release_path,
             &kernel_cmdline,
-            kernel_path,
-            initrd_path,
-            &esp_gen_paths,
+            &bootspec.kernel,
+            &kernel_target,
+            &initrd_location,
+            &initrd_target,
             &self.esp_paths.esp,
         )
         .context("Failed to assemble lanzaboote image.")?;
+        let stub_target = self
+            .esp_paths
+            .linux
+            .join(stub_name(generation, &self.key_pair.public_key)?);
+        self.gc_roots.extend([&stub_target]);
+        install_signed(&self.key_pair, &lanzaboote_image, &stub_target)
+            .context("Failed to install the Lanzaboote stub.")?;
+
+        Ok(())
+    }
+
+    /// Register the files of an already installed generation as garbage collection roots.
+    ///
+    /// An error should not be considered fatal; the generation should be (re-)installed instead.
+    fn register_installed_generation(&mut self, generation: &Generation) -> Result<()> {
```

Comment:
The name is not really descriptive of what this function does. In the control flow, it primarily checks if a generation is already installed and if it is, it registers the paths. 

```suggestion
    fn check_generation(&mut self, generation: &Generation) -> Result<()> {
```

Reply:
Hm, I thought of `register_already_installed_generation` at first, but that seemed a bit on the long side. `check_generation` is even less descriptive I think.

Reply:
Well the point is that the primary function of this function is not to register the generation but to check whether it is already installed and then skip installing the files.

Reply:
Depends on how you look at it. It is in fact a very important job of this function to register the generation, because otherwise the garbage collector would come and delete everything that was already installed.